platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongList.java

```diff
@@ -291,26 +390,109 @@ protected final void writeHeader(final FileChannel fc) throws IOException {
      * Lookup a long in data
      *
      * @param chunkIndex the index of the chunk the long is contained in
-     * @param subIndex The sub index of the long in that chunk
+     * @param subIndex   The sub index of the long in that chunk
      * @return The stored long value at given index
      */
     protected abstract long lookupInChunk(final long chunkIndex, final long subIndex);
 
     /**
-     * Current implementation is no-op.
+     * After invocation of this method, {@link LongList#get(long)}) calls
+     * will return {@link LongList#IMPERMISSIBLE_VALUE} for indices that
+     * are before {@code newMinValidIndex}.
+     * Also, a call to this method releases memory taken by unused byte buffers.
      *
      * @param newMinValidIndex minimal valid index of the list
      */
-    public void updateMinValidIndex(final long newMinValidIndex) {
-        // no-op
+    public final void updateMinValidIndex(final long newMinValidIndex) {
+        if (newMinValidIndex < 0) {
+            throw new IndexOutOfBoundsException("Min valid index " + newMinValidIndex + " must be non-negative");
+        }
+
+        if (newMinValidIndex >= size.get()) {
+            logger.warn("New min valid index cannot exceed current list size, returning.");
+            return;
+        }
+
+        // This is an optimization: we know that a chunk for this index must exist,
+        // so let's make sure that it's there
+        createOrGetChunk(newMinValidIndex);
+        shrinkIfNeeded(newMinValidIndex);
+        minValidIndex.set(newMinValidIndex);
```

Comment:
This line should probably be moved to before createOrGetChunk() and shrinkIfNeeded() to enable min index checks in put() in parallel threads.

Reply:
Yes. Probably you're right. I was trying to choose between these two options and thought that it may be better if we update it after all the preparations done. As far as I can see it, they outcome should be the same. There may be side effects in both cases but they are insignificant. In any case, I'll move it.